### DISCUSSION TITLE
proper way to set group label names for which-key only for norg files

### DISCUSSION THREAD

**Discussion by osamuaoki** [nvim-neorg/neorg#940] [Q&A] [answered]
Currently, norg doesn't set group label names so I see many "+prefix" displayed for non-final key steps in multi-key sequences.

I can't figure out how to bind  group label names display just for norg file buffers.

Let me explain with an example.  The following discussion example uses the setting:
```
config = {
      default_keybinds = true,
      neorg_leader = "<leader>n",
 },
```
Typing `<leader>` displays:
![Screenshot from 2023-06-09 17-20-31](https://github.com/nvim-neorg/neorg/assets/7339387/13f7eefc-28ef-4516-97a4-18ea9115725f)
Then typing `n` displays:
![Screenshot from 2023-06-09 17-20-47](https://github.com/nvim-neorg/neorg/assets/7339387/64a84042-b281-480f-95ea-ad9edd9ef667)

I know I can add proper group label names by tweaking which-key.  But this get activated even for non-norg files.
```
return {
  -- which-key
  {
    "folke/which-key.nvim",
    opts = {
      -- plugins = { spelling = true },
      defaults = {
        ["<leader>n"] = { name = "+norg" },
        ["<leader>ni"] = { name = "+insert" },
        ["<leader>nl"] = { name = "+list" },
        ["<leader>nm"] = { name = "+mode" },
        ["<leader>nn"] = { name = "+note" },
        ["<leader>nt"] = { name = "+task" },
      },
    },
  },
}
```
Slightly better (but hackish fix can be done in keybinds.lua of norg:
```local module = neorg.modules.create("core.keybinds.keybinds")
---@class core.keybinds
module.config.public = {
    keybind_presets = {
        neorg = function(keybinds)
            local leader = keybinds.leader

            -- Map all the below keybinds only when the "norg" mode is active
            -- I want to display "+neorg" in correct position as group label which are grouped after normal bind descriptions
            require("which-key").register({
                  [ leader ] = { name = "+neorg"},
                  [ leader .. "i" ] = { name = "+insert" },
                  [ leader .. "l" ] = { name = "+list" },
                  [ leader .. "m" ] = { name = "+mode" },
                  [ leader .. "n" ] = { name = "+note" },
                  [ leader .. "t" ] = { name = "+task" },
                  })
            keybinds.map_event_to_mode("norg", {
                n = {
                   ......
```
Then typing `<leader>` displays
![Screenshot from 2023-06-09 17-18-27](https://github.com/nvim-neorg/neorg/assets/7339387/0f684c64-4e96-4d00-9148-6b98bbba4f40)
Then typing `n` displays:
![Screenshot from 2023-06-09 17-17-52](https://github.com/nvim-neorg/neorg/assets/7339387/09de09e9-df6f-48fd-8fb4-0d5968213b2d)

This looks better but not really better than whick-key configuration above when multiple norg and no-neorg files are edited

I think `keybinds.map_event_to_mode("norg", ...` like assignment is needed for group label name too.  That part of code looked a bit too complicated for me to hack.

Anyone have thoughts?

(I don't think #381 nor #679 address issues discussed here.)

**Comment by max397574** [comment, marked as answer]
when you do that thing inside the neorg config that does absolutely nothing
the only thing it does it that the mappings aren't registered until you start neorg the first time
then they'll also be in non-norg buffers
group names is a which-key thing
this shouldn't be done from neorg

you can just do something like this:
```lua
vim.api.nvim_create_autocmd("BufEnter",{
pattern="*.norg",
callback=function()
wk.register(...,{buffer=true})
end
})
```

**Reply by osamuaoki** [reply]
Very true.  autocmd is the way to do this.  Thanks

**Reply by ddbrierton** [reply]
I’m pleased I stumbled across this as I was trying to solve exactly the same problem. Two things, though. (1) I think it needs to be `buffer = 0` not `buffer = true` in the opts table, the second argument to `wk.register()`. I have this:

```lua
vim.api.nvim_create_autocmd("BufEnter", {
  pattern = "*.norg",
  callback = function()
    require("which-key").register({
      ["<leader>n"] = { name = "+neorg" },
      ["<leader>ni"] = { name = "+insert" },
      ["<leader>nl"] = { name = "+list" },
      ["<leader>nm"] = { name = "+mode" },
      ["<leader>nn"] = { name = "+note" },
      ["<leader>nt"] = { name = "+task" },
    }, { buffer = 0 })
  end,
})
```

(2) Backspace no longer works to go back once I’m in insert/list/mode/note/task. No other which-key groups are affected, backspace continues to work as expected. So `<leader> n <bs>` works as expected, but `<leader> n i <bs>` leaves me stuck in the `+insert` group, and I have to either choose one of the options or `<esc>` to close which-keys completely.

Did you notice the same?

**Comment by weberam2** [comment]
OMG THANK YOU!
